content/2016-03-09-resample.rst

```diff
@@ -0,0 +1,167 @@
+        return results
+
+And then call the ``pythran`` compiler:
+
+.. code:: sh
+
+    % pythran resample.py
+
+This turns the Python file into a native extension, namely ``resample.so`` on Linux. Running it yields a nice speedup:
+
+.. code:: sh
+
+    % python -m timeit 'import numpy as np; np.random.seed(0) ; n = 1000; xs = np.arange(n, dtype=np.float64); qs = np.array([1.0/n,]*n); rands = np.random.rand(n); from resample import resample' 'resample(xs, qs, rands)'
+    1000 loops, best of 3: 1.23 msec per loop
+
+Second step: Pythran + OpenMP
+=============================
+
+But could we do better? An astute reader would note that the for loop can be
+run in parallel (iterations are independent). There's a famous standard for C,
+C++ and Fortran to parallelize this kind of trivial loops (and do many non
+trivial stuff also, but that's not the point here) called `OpenMP
+<http://openmp.org/>`_. It turns out Pythran supports OpenMP :-). By adding an extra comment (that should look pretty familiar to anyone accustomed to OpenMP) on the parallel loop:
+
+.. code:: python
+
+    #pythran export resample(float[], float[], float[])
+    import numpy as np
+    def resample(qs, xs, rands):
+        results = np.empty_like(qs)
+        lookup = np.cumsum(qs)
+        #omp parallel for
+        for j, key in enumerate(rands):
+            i = np.argmax(lookup > key)
+            results[j] = xs[i]
+        return results
+
+And adding the ``-fopenmp`` flag to the ``pythran`` call:
+
+.. code:: sh
+
+    % pythran resample.py -fopenmp
+
+We get an extra speedup (only two cores there, sorry about this :-/):
+
+.. code:: sh
+
+    % python -m timeit 'import numpy as np; np.random.seed(0) ; n = 1000; xs = np.arange(n, dtype=np.float64); qs = np.array([1.0/n,]*n); rands = np.random.rand(n); from resample import resample' 'resample(xs, qs, rands)'
+    1000 loops, best of 3: 693 usec per loop
+
+
+Third step: Pythran + Brain
+===========================
+
+Now wait… calling ``np.argmax`` on an array of ``bool`` is indeed a nice trick to get the index of the first value where ``lookup > key``, but it evaluates the whole array. There's no early exit, while there could be (there's only ``0`` and ``1`` after all). As pointed out on the `SO thread
+<http://stackoverflow.com/questions/21468170/numba-code-slower-than-pure-python>`_, one could write a ``np_index(array_expr)`` function that behaves like the ``list.index`` one:
+
+.. code:: python
+
+    #pythran export resample(float[], float[], float[])
+    import numpy as np
+
+    def np_index(haystack, needle):
+        for i, v in enumerate(haystack):
+            if v == needle:
+                return i
+        raise ValueError("Value not found")
+
+    def resample(qs, xs, rands):
+        results = np.empty_like(qs)
+        lookup = np.cumsum(qs)
+        for j, key in enumerate(rands):
+            i = np_index(lookup > key, True)
+            results[j] = xs[i]
+        return results
```

Comment:
> You didn't implement an overload of argmax for fexpr? :)

 The problem is not fexpr, but boolean. In fact for any type, if one
 value is the max possible value for this type then we could short
 circuit. It's trivial to implement and we are 100% sure it's benefical
 for bool. for other types, it adds an extra check and may not be
 profitable. I'll open an issue for this.